Update to PSR-11 for lazy listener support

weierophinney opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

This patch does the following:

- Bumps to PHP 7.4. This is necessary, as PSR-11 from 1.1 forward requires PHP 7.4.
- Switches from suggesting container-interop to suggesting PSR-11 directly, and allowing either v1 or v2 releases.
- Updates the various classes implementing lazy listeners to typehints against the PSR-11 `ContainerInterface` instead of the container-interop equivalent. This can be done because container-interop from v1.2 forward extended the PSR-11 interfaces; as such, this is type widening, which is allowed within PHP.

The impetus for this change is the [laminas-servicemanager 3.11.0 release](https://github.com/laminas/laminas-servicemanager/releases/tag/3.11.0), which now makes container-interop into class aliases of PSR-11.

weierophinney committed:
feat: bump dependencies

- Bumps to PHP 7.4, as psr/composer 1.1 series requires PHP 7.4
- Replaces container-interop requirement with PSR-11

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: switch to PSR-11 interface for typehints

Since container-interop from version 1.2 and up extends PSR-11, switching to PSR-11 for a typehint is technically type-widening, which is allowed for parameters in PHP.
As such, this is a backwards compatible change.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius commented:
Do we want to hold off until `psr/container:^2` actually works in `laminas/laminas-servicemanager`?

boesing commented on composer.json:
```diff
@@ -20,24 +20,27 @@
     },
     "suggest": {
-        "container-interop/container-interop": "^1.1, to use the lazy listeners feature",
+        "psr/container": "^1.1.2 || ^2.0.2, to use the lazy listeners feature",
```
I do not really like the constraint annotation. Lets take the `stdlib` as an example. We suggest using `stdlib` v2 but we do not even test it anymore (due to the fact it was somewhen removed).
Is this some kind of best practice or just done here to keep it in sync with `stdlib`?

boesing commented on composer.json:
```diff
@@ -20,24 +20,27 @@
     },
     "require-dev": {
-        "container-interop/container-interop": "^1.1",
+        "psr/container": "^1.1.2 || ^2.0.2",
```
shouldn't this be a requirement rather than a dev requirement?
The code is actively typing against and since these PSR standards are only interfaces, requiring them would be fine?

boesing reviewed (changes requested):
We should add conflict with `container-interop/container-interop` <1.2
I'd say that would not be necessary if we never supported interop but we did and thus, conflicting with <1.2 might be fine?
WDYT?

weierophinney commented:
> Do we want to hold off until psr/container:^2 actually works in laminas/laminas-servicemanager?

No, because we have both v1 and v2 support in this AND it's opt-in anyways.

Additionally they'll get a conflict if they try and install laminas-servicemanager AND psr/container v2 at the same time anyways. But once laminas-servicemanager supports psr/container v2, this will just work.

weierophinney committed:
fix: conflict with container-interop < 1.2

This ensures compatibility with current and future versions of laminas-servicemanager.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@boesing I've updated to conflict with container-interop < 1.2.
